File: androidApp/src/main/kotlin/app/hifis/hifis/schedule/EventFormSheet.kt
package app.hifis.hifis.schedule

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import app.hifis.hifis.R
import app.hifis.hifis.ui.tap
import app.hifis.hifis.ui.theme.Dimens
import app.hifis.hifis.ui.theme.HifisTheme
import app.hifis.hifis.ui.theme.HifisType
import app.hifis.hifis.ui.theme.eventColor
import app.hifis.shared.schedule.EventKind
import app.hifis.shared.schedule.EventPalette
import app.hifis.shared.schedule.EventScope

/**
 * 일정 추가 — 아래에서 올라오는 판
 *
 * 담는 것: 제목 · 시작 · 종료 · 종류 · 공유 범위 · 색 · 메모.
 *
 * **아직 아무것도 저장되지 않는다.** 서버가 없어서 `저장` 을 누르면 그냥 닫힌다.
 * 날짜·시각도 지금은 글자로 받는다 — 고르개는 다음이다.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun EventFormSheet(onDismiss: () -> Unit) {
    val colors = HifisTheme.colors
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    var title by remember { mutableStateOf("") }
    var startDate by remember { mutableStateOf("") }
    var startTime by remember { mutableStateOf("") }
    var endDate by remember { mutableStateOf("") }
    var endTime by remember { mutableStateOf("") }
    var kind by remember { mutableStateOf(EventKind.MEETING) }
    var scope by remember { mutableStateOf(EventScope.CENTER) }
    var colorIndex by remember { mutableIntStateOf(EventPalette.DEFAULT) }
    var memo by remember { mutableStateOf("") }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = colors.surface,
        dragHandle = {
            Box(Modifier.fillMaxWidth().padding(vertical = 10.dp), Alignment.Center) {
                Box(
                    Modifier
                        .size(width = 36.dp, height = 4.dp)
                        .background(colors.line, CircleShape),
                )
            }
        },
    ) {
        Column(
            Modifier
                .verticalScroll(rememberScrollState())
                .padding(horizontal = Dimens.screenEdge)
                .padding(bottom = 32.dp),
        ) {
            SheetHeader(onDismiss)

            Section("제목")
            Field(title, { title = it }, "무엇을 계획하고 있나요?")

            Section("시작")
            DateTimeRow(startDate, { startDate = it }, startTime, { startTime = it })

            Section("종료")
            DateTimeRow(endDate, { endDate = it }, endTime, { endTime = it })

            Section("카테고리")
            KindChips(kind) { kind = it }

            Section("공유 범위")
            ScopeCards(scope) { scope = it }

            Section("색상")
            ColorSwatches(colorIndex) { colorIndex = it }

            Section("메모", optional = true)
            Field(
                memo,
                { memo = it },
                "참석자·장소·준비물 등 상세 내용을 적어주세요",
                minHeight = 110.dp,
            )

            Spacer(Modifier.height(28.dp))
            SaveButton(enabled = title.isNotBlank(), onClick = onDismiss)
        }
    }
}

@Composable
private fun SheetHeader(onClose: () -> Unit) {
    val colors = HifisTheme.colors
    Row(verticalAlignment = Alignment.CenterVertically) {
        Box(
            Modifier
                .size(48.dp)
                .background(
                    colors.brand.copy(alpha = if (colors.isDark) 0.20f else 0.12f),
                    RoundedCornerShape(14.dp),
                ),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                painterResource(R.drawable.ic_schedule),
                contentDescription = null,
                tint = colors.brand,
                modifier = Modifier.size(24.dp),
            )
        }
        Spacer(Modifier.width(14.dp))
        Column(Modifier.weight(1f)) {
            Text("일정 추가", style = HifisType.title.copy(fontSize = 20.sp()), color = colors.ink)
            Text(
                "센터와 공유할 일정을 만들어요",
                style = HifisType.caption,
                color = colors.inkTertiary,
            )
        }
        Box(
            Modifier
                .size(36.dp)
                .clip(CircleShape)
                .tap(label = "닫기", onClick = onClose),
            contentAlignment = Alignment.Center,
        ) {
            Icon(
                painterResource(R.drawable.ic_close),
                contentDescription = "닫기",
                tint = colors.inkTertiary,
                modifier = Modifier.size(20.dp),
            )
        }
    }
}

/** `20.sp` 를 스타일 안에서 바꾸려고 쓰는 잔손질 */
private fun Int.sp() = androidx.compose.ui.unit.TextUnit(
    this.toFloat(),
    androidx.compose.ui.unit.TextUnitType.Sp,
)

@Composable
private fun Section(label: String, optional: Boolean = false) {
    val colors = HifisTheme.colors
    Spacer(Modifier.height(22.dp))
    Row(verticalAlignment = Alignment.Bottom) {
        Text(
            label,
            style = HifisType.label.copy(fontWeight = FontWeight.Bold),
            color = colors.ink,
        )
        if (optional) {
            Spacer(Modifier.width(5.dp))
            Text("(선택)", style = HifisType.caption, color = colors.inkTertiary)
        }
    }
    Spacer(Modifier.height(10.dp))
}

/** 글자를 받는 칸 — 면은 [HifisColors.fieldFill] 이다 */
@Composable
private fun Field(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    modifier: Modifier = Modifier,
    trailingIcon: Int? = null,
    minHeight: androidx.compose.ui.unit.Dp = 52.dp,
) {
    val colors = HifisTheme.colors
    val shape = RoundedCornerShape(14.dp)
    BasicTextField(
        value = value,
        onValueChange = onValueChange,
        textStyle = HifisType.label.copy(color = colors.ink),
        cursorBrush = SolidColor(colors.brand),
        modifier = modifier
            .fillMaxWidth()
            .heightIn(min = minHeight)
            .background(colors.fieldFill, shape),
        decorationBox = { inner ->
            Row(
                Modifier.padding(horizontal = 16.dp, vertical = 15.dp),
                verticalAlignment = if (minHeight > 60.dp) Alignment.Top else Alignment.CenterVertically,
            ) {
                Box(Modifier.weight(1f)) {
                    if (value.isEmpty()) {
                        Text(placeholder, style = HifisType.label, color = colors.inkTertiary)
                    }
                    inner()
                }
                if (trailingIcon != null) {
                    Spacer(Modifier.width(10.dp))
                    Icon(
                        painterResource(trailingIcon),
                        contentDescription = null,
                        tint = colors.inkTertiary,
                        modifier = Modifier.size(18.dp),
                    )
                }
            }
        },
    )
}

/** 날짜 + 시각 — 날짜가 더 길어서 자리를 더 준다 */
@Composable
private fun DateTimeRow(
    date: String,
    onDate: (String) -> Unit,
    time: String,
    onTime: (String) -> Unit,
) {
    Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
        Field(date, onDate, "YYYY-MM-DD", Modifier.weight(1.7f), R.drawable.ic_schedule)
        Field(time, onTime, "--:--", Modifier.weight(1f), R.drawable.ic_attendance)
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun KindChips(picked: EventKind, onPick: (EventKind) -> Unit) {
    val colors = HifisTheme.colors
    FlowRow(
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        EventKind.all.forEach { kind ->
            val on = kind == picked
            Row(
                Modifier
                    .clip(CircleShape)
                    .background(
                        if (on) colors.brand.copy(alpha = if (colors.isDark) 0.18f else 0.10f)
                        else Color.Transparent,
                        CircleShape,
                    )
                    .border(1.dp, if (on) colors.brand else colors.line, CircleShape)
                    .tap(label = kind.label) { onPick(kind) }
                    .padding(horizontal = 14.dp, vertical = 10.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Icon(
                    painterResource(drawableOf(kind.icon)),
                    contentDescription = null,
                    tint = if (on) colors.brand else colors.inkSecondary,
                    modifier = Modifier.size(16.dp),
                )
                Spacer(Modifier.width(6.dp))
                Text(
                    kind.label,
                    style = HifisType.caption.copy(fontWeight = FontWeight.SemiBold),
                    color = if (on) colors.brand else colors.ink,
                )
            }
        }
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun ScopeCards(picked: EventScope, onPick: (EventScope) -> Unit) {
    val colors = HifisTheme.colors
    val shape = RoundedCornerShape(14.dp)
    FlowRow(
        maxItemsInEachRow = 2,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        EventScope.all.forEach { scope ->
            val on = scope == picked
            Column(
                Modifier
                    .weight(1f)
                    .clip(shape)
                    .background(
                        if (on) colors.brand.copy(alpha = if (colors.isDark) 0.16f else 0.08f)
                        else Color.Transparent,
                        shape,
                    )
                    .border(1.dp, if (on) colors.brand else colors.line, shape)
                    .tap(label = scope.label) { onPick(scope) }
                    .padding(horizontal = 14.dp, vertical = 13.dp),
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        painterResource(drawableOf(scope.icon)),
                        contentDescription = null,
                        tint = if (on) colors.brand else colors.inkSecondary,
                        modifier = Modifier.size(17.dp),
                    )
                    Spacer(Modifier.width(7.dp))
                    Text(
                        scope.label,
                        style = HifisType.label.copy(fontWeight = FontWeight.Bold),
                        color = if (on) colors.brand else colors.ink,
                    )
                }
                Spacer(Modifier.height(3.dp))
                Text(scope.detail, style = HifisType.caption, color = colors.inkTertiary)
            }
        }
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun ColorSwatches(picked: Int, onPick: (Int) -> Unit) {
    val colors = HifisTheme.colors
    FlowRow(
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        EventPalette.colors.indices.forEach { index ->
            val on = index == picked
            Box(
                Modifier
                    .size(34.dp)
                    .clip(CircleShape)
                    .background(eventColor(index), CircleShape)
                    // 고른 것에만 바탕색 테를 둘러 띄운다 — 색끼리 붙어 있어 테가 없으면 못 찾는다
                    .then(if (on) Modifier.border(2.5.dp, colors.ink, CircleShape) else Modifier)
                    .tap(label = "색 ${index + 1}") { onPick(index) },
                contentAlignment = Alignment.Center,
            ) {
                if (on) {
                    Icon(
                        painterResource(R.drawable.ic_check),
                        contentDescription = null,
                        tint = Color.White,
                        modifier = Modifier.size(16.dp),
                    )
                }
            }
        }
    }
}

@Composable
private fun SaveButton(enabled: Boolean, onClick: () -> Unit) {
    val colors = HifisTheme.colors
    val shape = RoundedCornerShape(14.dp)
    Box(
        Modifier
            .fillMaxWidth()
            .height(54.dp)
            .clip(shape)
            .background(if (enabled) colors.brand else colors.fieldFill, shape)
            .then(if (enabled) Modifier.tap(label = "저장", onClick = onClick) else Modifier),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            "저장",
            style = HifisType.body.copy(fontWeight = FontWeight.Bold),
            color = if (enabled) Color.White else colors.inkTertiary,
        )
    }
}

/**
 * 아이콘 이름을 그림 자원으로 — **`shared` 가 이름만 들고 있어서 여기서 잇는다**
 *
 * 여기 빠진 이름이 있으면 그 칩만 그림이 없다. `shared` 의 테스트가
 * 이름이 `ic_` 로 시작하는지까지는 보지만, 자원이 있는지는 이 표가 지킨다.
 */
private fun drawableOf(icon: String): Int = when (icon) {
    "ic_people" -> R.drawable.ic_people
    "ic_dumbbell" -> R.drawable.ic_dumbbell
    "ic_attendance" -> R.drawable.ic_attendance
    "ic_pin" -> R.drawable.ic_pin
    "ic_sun" -> R.drawable.ic_sun
    "ic_moon" -> R.drawable.ic_moon
    "ic_notice" -> R.drawable.ic_notice
    "ic_gift" -> R.drawable.ic_gift
    "ic_work" -> R.drawable.ic_work
    "ic_person_check" -> R.drawable.ic_person_check
    "ic_cap" -> R.drawable.ic_cap
    "ic_chat" -> R.drawable.ic_chat
    "ic_glass" -> R.drawable.ic_glass
    "ic_heart" -> R.drawable.ic_heart
    "ic_person" -> R.drawable.ic_person
    "ic_dots" -> R.drawable.ic_dots
    "ic_more" -> R.drawable.ic_more
    "ic_branch" -> R.drawable.ic_branch
    "ic_project" -> R.drawable.ic_project
    else -> R.drawable.ic_dots
}

File: iosApp/iosApp/Schedule/EventFormSheet.swift
import SwiftUI
import SharedKit

/// 일정 추가 — 아래에서 올라오는 판
///
/// 담는 것: 제목 · 시작 · 종료 · 종류 · 공유 범위 · 색 · 메모.
/// 안드로이드 `EventFormSheet.kt` 와 같은 판이다 — 한쪽만 고치면 갈린다.
///
/// **아직 아무것도 저장되지 않는다.** 서버가 없어서 `저장` 을 누르면 그냥 닫힌다.
/// 날짜·시각도 지금은 글자로 받는다 — 고르개는 다음이다.
struct EventFormSheet: View {
    let onDismiss: () -> Void

    @State private var title = ""
    @State private var startDate = ""
    @State private var startTime = ""
    @State private var endDate = ""
    @State private var endTime = ""
    @State private var kind = EventKind.meeting
    @State private var scope = EventScope.center
    @State private var colorIndex: Int32 = Int32(EventPalette.shared.DEFAULT)
    @State private var memo = ""

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                header

                section("제목")
                field($title, "무엇을 계획하고 있나요?")

                section("시작")
                dateTimeRow($startDate, $startTime)

                section("종료")
                dateTimeRow($endDate, $endTime)

                section("카테고리")
                kindChips

                section("공유 범위")
                scopeCards

                section("색상")
                colorSwatches

                section("메모", optional: true)
                field($memo, "참석자·장소·준비물 등 상세 내용을 적어주세요", minHeight: 110)

                Spacer().frame(height: 28)
                saveButton
            }
            .padding(.horizontal, HifisSize.screenEdge)
            .padding(.bottom, 32)
        }
        .background(HifisColor.surface)
        .presentationDragIndicator(.visible)
    }

    private var header: some View {
        HStack(spacing: 14) {
            Image("ic_schedule")
                .renderingMode(.template)
                .resizable()
                .frame(width: 24, height: 24)
                .foregroundStyle(HifisColor.brand)
                .frame(width: 48, height: 48)
                .background(
                    RoundedRectangle(cornerRadius: 14, style: .continuous)
                        .fill(HifisColor.brand.opacity(HifisColor.toneFillOpacity))
                )
            VStack(alignment: .leading, spacing: 1) {
                Text("일정 추가")
                    .font(.system(size: 20, weight: .bold))
                    .foregroundStyle(HifisColor.ink)
                Text("센터와 공유할 일정을 만들어요")
                    .font(HifisFont.caption)
                    .foregroundStyle(HifisColor.inkTertiary)
            }
            Spacer(minLength: 8)
            Button(action: onDismiss) {
                Image("ic_close")
                    .renderingMode(.template)
                    .resizable()
                    .frame(width: 20, height: 20)
                    .foregroundStyle(HifisColor.inkTertiary)
                    .frame(width: 36, height: 36)
                    .contentShape(Rectangle())
            }
            .buttonStyle(TapStyle())
            .accessibilityLabel("닫기")
        }
        .padding(.top, 8)
    }

    private func section(_ label: String, optional: Bool = false) -> some View {
        HStack(alignment: .bottom, spacing: 5) {
            Text(label)
                .font(.system(size: 14, weight: .bold))
                .foregroundStyle(HifisColor.ink)
            if optional {
                Text("(선택)")
                    .font(HifisFont.caption)
                    .foregroundStyle(HifisColor.inkTertiary)
            }
        }
        .padding(.top, 22)
        .padding(.bottom, 10)
    }

    /// 글자를 받는 칸 — 면은 `fieldFill` 이다
    private func field(
        _ text: Binding<String>,
        _ placeholder: String,
        trailing: String? = nil,
        minHeight: CGFloat = 52
    ) -> some View {
        HStack(alignment: minHeight > 60 ? .top : .center, spacing: 10) {
            TextField("", text: text, axis: minHeight > 60 ? .vertical : .horizontal)
                .font(HifisFont.label)
                .foregroundStyle(HifisColor.ink)
                .tint(HifisColor.brand)
                .lineLimit(minHeight > 60 ? 4 : 1, reservesSpace: minHeight > 60)
                .overlay(alignment: .leading) {
                    if text.wrappedValue.isEmpty {
                        Text(placeholder)
                            .font(HifisFont.label)
                            .foregroundStyle(HifisColor.inkTertiary)
                            .allowsHitTesting(false)
                    }
                }
            if let trailing {
                Image(trailing)
                    .renderingMode(.template)
                    .resizable()
                    .frame(width: 18, height: 18)
                    .foregroundStyle(HifisColor.inkTertiary)
            }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 15)
        .frame(minHeight: minHeight, alignment: .topLeading)
        .background(
            HifisColor.fieldFill,
            in: RoundedRectangle(cornerRadius: 14, style: .continuous)
        )
    }

    /// 날짜 + 시각 — 날짜가 더 길어서 자리를 더 준다
    private func dateTimeRow(_ date: Binding<String>, _ time: Binding<String>) -> some View {
        HStack(spacing: 10) {
            field(date, "YYYY-MM-DD", trailing: "ic_schedule")
                .frame(maxWidth: .infinity)
            field(time, "--:--", trailing: "ic_attendance")
                .frame(width: 130)
        }
    }

    private var kindChips: some View {
        FlowLayout(spacing: 8, lineSpacing: 8) {
            ForEach(EventKind.companion.all, id: \.label) { item in
                let on = item == kind
                Button { kind = item } label: {
                    HStack(spacing: 6) {
                        Image(item.icon)
                            .renderingMode(.template)
                            .resizable()
                            .frame(width: 16, height: 16)
                            .foregroundStyle(on ? HifisColor.brand : HifisColor.inkSecondary)
                        Text(item.label)
                            .font(.system(size: 13, weight: .semibold))
                            .foregroundStyle(on ? HifisColor.brand : HifisColor.ink)
                    }
                    .padding(.horizontal, 14)
                    .padding(.vertical, 10)
                    .background(
                        Capsule().fill(on ? HifisColor.brand.opacity(0.18) : .clear)
                    )
                    .overlay(
                        Capsule().strokeBorder(on ? HifisColor.brand : HifisColor.line, lineWidth: 1)
                    )
                    .contentShape(Capsule())
                }
                .buttonStyle(TapStyle())
            }
        }
    }

    private var scopeCards: some View {
        LazyVGrid(
            columns: [GridItem(.flexible(), spacing: 10), GridItem(.flexible(), spacing: 10)],
            spacing: 10
        ) {
            ForEach(EventScope.companion.all, id: \.label) { item in
                let on = item == scope
                Button { scope = item } label: {
                    VStack(alignment: .leading, spacing: 3) {
                        HStack(spacing: 7) {
                            Image(item.icon)
                                .renderingMode(.template)
                                .resizable()
                                .frame(width: 17, height: 17)
                                .foregroundStyle(on ? HifisColor.brand : HifisColor.inkSecondary)
                            Text(item.label)
                                .font(.system(size: 14, weight: .bold))
                                .foregroundStyle(on ? HifisColor.brand : HifisColor.ink)
                        }
                        Text(item.detail)
                            .font(HifisFont.caption)
                            .foregroundStyle(HifisColor.inkTertiary)
                            .multilineTextAlignment(.leading)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .padding(.horizontal, 14)
                    .padding(.vertical, 13)
                    .background(
                        RoundedRectangle(cornerRadius: 14, style: .continuous)
                            .fill(on ? HifisColor.brand.opacity(0.16) : .clear)
                    )
                    .overlay(
                        RoundedRectangle(cornerRadius: 14, style: .continuous)
                            .strokeBorder(on ? HifisColor.brand : HifisColor.line, lineWidth: 1)
                    )
                    .contentShape(Rectangle())
                }
                .buttonStyle(TapStyle())
            }
        }
    }

    private var colorSwatches: some View {
        FlowLayout(spacing: 10, lineSpacing: 10) {
            ForEach(0..<Int(EventPalette.shared.colors.count), id: \.self) { index in
                let on = Int32(index) == colorIndex
                Button { colorIndex = Int32(index) } label: {
                    Circle()
                        .fill(HifisEventColor.at(Int32(index)))
                        .frame(width: 34, height: 34)
                        // 고른 것에만 테를 둘러 띄운다 — 색끼리 붙어 있어 테가 없으면 못 찾는다
                        .overlay(
                            Circle().strokeBorder(on ? HifisColor.ink : .clear, lineWidth: 2.5)
                        )
                        .overlay {
                            if on {
                                Image("ic_check")
                                    .renderingMode(.template)
                                    .resizable()
                                    .frame(width: 16, height: 16)
                                    .foregroundStyle(.white)
                            }
                        }
                }
                .buttonStyle(TapStyle())
                .accessibilityLabel("색 \(index + 1)")
            }
        }
    }

    private var saveButton: some View {
        let enabled = !title.trimmingCharacters(in: .whitespaces).isEmpty
        return Button { if enabled { onDismiss() } } label: {
            Text("저장")
                .font(.system(size: 16, weight: .bold))
                .foregroundStyle(enabled ? .white : HifisColor.inkTertiary)
                .frame(maxWidth: .infinity)
                .frame(height: 54)
                .background(
                    RoundedRectangle(cornerRadius: 14, style: .continuous)
                        .fill(enabled ? HifisColor.brand : HifisColor.fieldFill)
                )
                .contentShape(Rectangle())
        }
        .buttonStyle(TapStyle())
        .disabled(!enabled)
    }
}

/// 칩처럼 폭이 제각각인 것을 줄바꿈하며 세운다
///
/// SwiftUI 에는 이런 배치가 없다 (`LazyVGrid` 는 칸 폭이 같아야 한다).
/// 안드로이드의 `FlowRow` 자리다.
struct FlowLayout: Layout {
    var spacing: CGFloat = 8
    var lineSpacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let maxWidth = proposal.width ?? .infinity
        var x: CGFloat = 0, y: CGFloat = 0, lineHeight: CGFloat = 0
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x > 0, x + size.width > maxWidth {
                x = 0
                y += lineHeight + lineSpacing
                lineHeight = 0
            }
            x += size.width + spacing
            lineHeight = max(lineHeight, size.height)
        }
        return CGSize(width: maxWidth == .infinity ? x : maxWidth, height: y + lineHeight)
    }

    func placeSubviews(
        in bounds: CGRect,
        proposal: ProposedViewSize,
        subviews: Subviews,
        cache: inout ()
    ) {
        var x: CGFloat = 0, y: CGFloat = 0, lineHeight: CGFloat = 0
        for view in subviews {
            let size = view.sizeThatFits(.unspecified)
            if x > 0, x + size.width > bounds.width {
                x = 0
                y += lineHeight + lineSpacing
                lineHeight = 0
            }
            view.place(
                at: CGPoint(x: bounds.minX + x, y: bounds.minY + y),
                proposal: ProposedViewSize(size)
            )
            x += size.width + spacing
            lineHeight = max(lineHeight, size.height)
        }
    }
}
